x/mongo/driver/topology/server.go

```diff
@@ -617,6 +617,10 @@ func (s *Server) setupHeartbeatConnection() error {
 
 	// Take the lock when assigning the context and connection because they're accessed by cancelCheck.
 	s.heartbeatLock.Lock()
+	if s.heartbeatCtxCancel != nil {
+		// Ensure the previous context is cancelled to avoid a leak.
+		s.heartbeatCtxCancel()
+	}
```

Comment:
To check my understanding. I think this bug would manifest as one context leaked per downed server every `heartbeatFrequency`. So, it would be a very slow leak in practice. Right?

Reply:
That's correct. In addition, each context value is very small because it's only holding a cancellation signal and no other values. The leak would be much bigger if this were a user-provided context, which could potentially hold a lot of memory due to `context.WithValue` calls.

To provided some perspective on the leak speed: I tried to benchmark the leak by setting internally setting `heartbeatFrequency` to 10ms and sleeping for 10 seconds, thereby causing ~1000 failed connection attempts in the heartbeat goroutine and consequently ~1000 leaked contexts, but that wasn't enough to get this to show up in the top 10 list of heap allocations.

Reply:
Oh that is a very slow leak. Thanks for the explanation!